src/OAuth/Server.php

```diff
@@ -0,0 +1,180 @@
+        $this->auth_server->enableGrantType(
+            new ClientCredentialsGrant(),
+            new \DateInterval(GLPI_OAUTH_ACCESS_TOKEN_EXPIRES)
+        );
+
+        $this->auth_server->enableGrantType(
+            new PasswordGrant(
+                new UserRepository(),
+                new RefreshTokenRepository()
+            ),
+            new \DateInterval(GLPI_OAUTH_ACCESS_TOKEN_EXPIRES)
+        );
+
+        $this->auth_server->enableGrantType(
+            new AuthCodeGrant(
+                new AuthCodeRepository(),
+                new RefreshTokenRepository(),
+                new \DateInterval('PT10M')
+            ),
+            new \DateInterval(GLPI_OAUTH_ACCESS_TOKEN_EXPIRES)
+        );
+
+        $this->auth_server->enableGrantType(
+            new RefreshTokenGrant(new RefreshTokenRepository()),
+            new \DateInterval(GLPI_OAUTH_ACCESS_TOKEN_EXPIRES)
+        );
+    }
+
+    private static function getInstance(): self
+    {
+        static $instance = null;
+        if ($instance === null) {
+            $instance = new self();
+        }
+        return $instance;
+    }
+
+    public static function getAuthorizationServer(): AuthorizationServer
+    {
+        return (self::getInstance())->auth_server;
+    }
+
+    /**
+     * @param Request $request
+     * @return array
+     * @phpstan-return {client_id: string, user_id: string, scopes: string[]}
+     * @throws \League\OAuth2\Server\Exception\OAuthServerException
+     */
+    public static function validateAccessToken(Request $request): array
+    {
+        $new_request = self::getInstance()->resource_server->validateAuthenticatedRequest($request);
+        return [
+            'client_id' => $new_request->getAttribute('oauth_client_id'),
+            'user_id' => $new_request->getAttribute('oauth_user_id'),
+            'scopes' => $new_request->getAttribute('oauth_scopes'),
+        ];
+    }
+
+    /**
+     * Get all scopes available for clients.
+     * @return array
+     */
+    public static function getAllowedScopes(): array
+    {
+        // Not yet supported
+        return [];
+    }
+
+    public static function generateKeys(): void
+    {
+        $private_key_path = GLPI_CONFIG_DIR . '/oauth.pem';
+        $public_key_path = GLPI_CONFIG_DIR . '/oauth.pub';
+        if (!file_exists($private_key_path) && !file_exists($public_key_path)) {
+            $config = [
+                'digest_alg'       => 'sha512',
+                'private_key_bits' => 2048,
+                'private_key_type' => OPENSSL_KEYTYPE_RSA,
+            ];
+            $res = openssl_pkey_new($config);
+            openssl_pkey_export_to_file($res, $private_key_path);
+            // Export public key to the public key file
+            $pubkey = openssl_pkey_get_details($res);
+            file_put_contents($public_key_path, $pubkey['key']);
+
+            chmod($private_key_path, 0660);
+            chmod($public_key_path, 0660);
```

Comment:
Code must be as defensive as possible here. Many methods can return `false` if something went wrong, and this should probably trigger an exception or make this method return false.